.circleci/bin/deploy-ci.py

```diff
@@ -0,0 +1,498 @@
+#!/usr/bin/env python3
+
+import json
+import logging
+import time
+from pprint import pprint
+from subprocess import PIPE, Popen
+from threading import Thread
+
+import click
+import requests
+
+logging.basicConfig(
+    format="%(levelname)-8s [%(asctime)s] %(message)s", level=logging.INFO
+)
+logger = logging.getLogger("cli")
+
+
+ENVIRONMENTS = ("staging",)
+SERVICES = ("all", "discovery", "creator", "identity")
+CREATOR_NODES = (
+    "stage-creator-4",  # canary
+    "stage-creator-5",
+    "stage-creator-6",
+    "stage-creator-7",
+    "stage-creator-8",
+    "stage-creator-9",
+    "stage-creator-10",
+    "stage-creator-11",
+    "stage-user-metadata",
+)
+DISCOVERY_NODES = (
+    "stage-discovery-1",
+    "stage-discovery-2",
+    "stage-discovery-3",
+    "stage-discovery-4",  # canary
+    "stage-discovery-5",
+)
+IDENTITY_NODES = ("stage-identity",)
+ALL_NODES = CREATOR_NODES + DISCOVERY_NODES + IDENTITY_NODES
+
+MASTER = "master"
+MISSING = "missing"
+RELEASE = "release"
+PRE_DEPLOY = "pre-deploy"
+POST_DEPLOY = "post-deploy"
+
+RAISE = "raise"
+EXIT_1 = "exit-1"
+IGNORE = "ignore"
+
+
+def get_service_from_host(host):
+    """Helper to detect the audius-cli service target."""
+
+    if host in CREATOR_NODES:
+        return "creator-node"
+    if host in DISCOVERY_NODES:
+        return "discovery-provider"
+    if host in IDENTITY_NODES:
+        return "identity-service"
+
+
+def ssh(host, cmd, exit_on_error=RAISE, show_output=False, dry_run=False):
+    """Helper for `run_cmd()` designed for running commands on a host via SSH."""
+
+    ssh_cmd = f"ssh -o LogLevel=quiet {host} -- {cmd}"
+    if dry_run:
+        logger.info(f"Dry run: {ssh_cmd}")
+    else:
+        return run_cmd(
+            ssh_cmd,
+            msg=f">> {host}: {cmd}",
+            exit_on_error=exit_on_error,
+            show_output=show_output,
+        )
+
+
+def run_cmd(cmd, exit_on_error=RAISE, msg=None, show_output=False):
+    """
+    Execute a shell command and return stdout.
+    Exit, raise error, or ignore on stderr.
+    """
+
+    # set the appropriate log level
+    log = logger.debug
+    if show_output:
+        log = logger.info
+
+    # run command and grab stdout/stderr
+    log(msg if msg else f"< {cmd}")
+    sp = Popen(cmd.split(" "), stdout=PIPE, stderr=PIPE)
+    stdout, stderr = sp.communicate()
+    stdout = stdout.strip().decode()
+    stderr = stderr.strip().decode()
+
+    # log stdout
+    if stdout:
+        log(stdout)
+
+    # log stderr
+    # or exit(1), raise error, or pass
+    if stderr:
+        logger.warning(stderr)
+        if "Could not get object for" in stderr:
+            log("REASON: Branch was deleted.")
+        if exit_on_error == RAISE:
+            raise RuntimeError("Previous command had stderr output.")
+        elif exit_on_error == EXIT_1:
+            exit(1)
+        else:
+            pass
```

Comment:
worth an "unknown" log here?

Reply:
It is known, it's IGNORE on line 50 and can only be one of the three: RAISE, EXIT_1, or IGNORE, as described on line 102.

The log I believe your looking for is present in line 104 under a WARNING loglevel.

I thought about doing an `elif exit_on_error == IGNORE: pass`, but figured it'd be best to be less restrictive and lean more towards duck-typing. Thoughts?

Reply:
sounds good 👍